src/Components/Components/src/NavigationManager.cs

```diff
@@ -274,6 +281,111 @@ protected void NotifyLocationChanged(bool isInterceptedLink)
         }
     }
 
+    /// <summary>
+    /// Notifies the registered handlers of the current location change.
+    /// </summary>
+    /// <param name="uri">The destination URI. This can be absolute, or relative to the base URI.</param>
+    /// <param name="isNavigationIntercepted">Whether this navigation was intercepted from a link.</param>
+    /// <returns>A <see cref="ValueTask{TResult}"/> representing the completion of the operation. If the result is <see langword="true"/>, the navigation should continue.</returns>
+    protected async ValueTask<bool> NotifyLocationChangingAsync(string uri, bool isNavigationIntercepted)
+    {
+        _locationChangingCts?.Cancel();
+        _locationChangingCts = null;
+
+        var handlerCount = _locationChangingHandlers.Count;
+
+        if (handlerCount == 0)
+        {
+            return true;
+        }
+
+        _locationChangingCts = new();
+
+        var context = new LocationChangingContext(uri, isNavigationIntercepted, _locationChangingCts);
+        var cancellationToken = _locationChangingCts.Token;
+
+        try
+        {
+            if (handlerCount == 1)
+            {
+                await _locationChangingHandlers[0].Invoke(context);
```

Comment:
Looks like the `handlerCount == 1` case has slightly different cancellation behavior than then `handlerCount != 1` case.

On this branch, we're not short-circuiting the `await` if the cancellation token fires, whereas on the other branch we do.

I don't fully know what difference it will make to the upstream caller. But should we bring these two branches into alignment?